### ISSUE TITLE
NUD Bounding Sphere implementation

### ISSUE THREAD

**Issue by Dr-HyperCake** [jam1garner/Smash-Forge#105]
At the start of each nud file (at 0x20), there are four floats, and at the start of each mesh definition there are eight floats. Smash Forge currently reads these in as unknowns, and writes them as 0 for imported models. However, after looking into it I think they are bounding spheres.

Here's how I think it works. A bounding sphere is four floats: x, y and z for the position of its center, and a radius. A nud has a global bounding sphere, which encompasses all of its geometry. Each mesh has two bounding spheres: the first encompasses the geometry of all polygons of the mesh, and the second, most of the time, has the x, y and z of the first but with a radius of zero.

The game uses the bounding sphere for depth testing, for the purposes of alpha blending correctly. In cases that a mesh is bone-weighted, or unbound, the bounding sphere is not used - the game tests the geometry itself for depth (presumably, at all times for weighted meshes, and at initalization for unbound ones).

However, for meshes that are single-bound to a bone, the game uses the bounding sphere to efficiently test for depth at all times. When the bounding sphere is just set to 0 (as Smash Forge does), single-bound meshes do not blend in the correct order relative to other meshes and models, because their bounding sphere is incorrect.

I have done some tests with the bounding sphere, and I have successfully set bounding spheres in such a way that meshes of the same model blend with each other correctly. However, I haven't successfully set the global bounding sphere of a nud in such a way that it blends properly with other nuds. I also have not yet figured out the purpose of a mesh's second bounding sphere, but in the majority of cases it is set as I mentioned.

In conclusion, I think Smash Forge should calculate bounding spheres for geometry, to be written when saving the nud file. To reiterate, it is:
- A bounding sphere for the nud as a whole, which is at 0x20 to 0x30 in each nud.
- Two bounding spheres for each mesh, which are at 0x0 to 0x10 and 0x10 to 0x20, respectively, for each mesh definition. The second bounding sphere is the same as the first except its radius should be 0.

**Comment by Dr-HyperCake** [comment]
I had not realized at the time of writing that Smash Forge was using the data as a bounding box already, only afterwards.

I have tested the newly added generation for bounding boxes, but it gives somewhat of an inferior result to what I would get from a bounding sphere calculation; where the sphere calculation would at least allow two blending meshes to render in the correct order, the Smash Forge calculations appear to have little effect on the blending as if it were still 0,0,0,0. Also, in cases of a nud with only one mesh, it does not generate the same values for the global bounding box as it does for the mesh's, when, logically, it should (and vanilla nud files have it as such). Due to the not-quite-accurate calculations, Smash Forge replaces the valid values in vanilla nud files when they are resaved.

Nonetheless, I appreciate the acknowledgement of the issue and implementation of a generation. I believe, though, that the calculation needs to be refined sometime in the future.
